apis/Google.Storage.V1/Google.Storage.V1.IntegrationTests/SignedUrlUtilsTest.cs

```diff
@@ -0,0 +1,487 @@
+                Method = HttpMethod.Put,
+                Headers = {
+                    { "x-goog-encryption-algorithm", "AES256" },
+                    { "x-goog-encryption-key", key },
+                    { "x-goog-encryption-key-sha256", hash }
+                }
+            };
+            putRequest.RequestUri = new Uri(SignedUrlUtils.Create(
+                bucket,
+                name,
+                _credentialsFilePath,
+                expiration: null,
+                request: putRequest));
+            var response = await _httpClient.SendAsync(putRequest);
+            Assert.True(response.IsSuccessStatusCode);
+
+            // Make sure the encryption succeeded.
+            await Assert.ThrowsAsync<GoogleApiException>(
+                () => _fixture.Client.DownloadObjectAsync(bucket, name, new MemoryStream()));
+            
+            Func<HttpRequestMessage> createGetRequest = () => new HttpRequestMessage()
+            {
+                Method = HttpMethod.Get,
+                Headers = {
+                    { "x-goog-encryption-algorithm", "AES256" },
+                    { "x-goog-encryption-key", key },
+                    { "x-goog-encryption-key-sha256", hash }
+                }
+            };
+            var request = createGetRequest();
+            var url = SignedUrlUtils.Create(
+                bucket,
+                name,
+                _credentialsFilePath,
+                duration,
+                request);
+            request.RequestUri = new Uri(url);
+
+            // Verify that the URL works initially.
+            response = await _httpClient.SendAsync(request);
+            var result = await response.Content.ReadAsByteArrayAsync();
+            Assert.Equal(data, result);
+
+            // Wait until the URL expires.
+            await Task.Delay(duration);
+
+            // Verify that the URL no longer works.
+            request = createGetRequest();
+            request.RequestUri = new Uri(url);
+            response = await _httpClient.SendAsync(request);
+            Assert.Equal(HttpStatusCode.BadRequest, response.StatusCode);
+
+            // Cleanup
+            await _fixture.Client.DeleteObjectAsync(bucket, name);
+        }
+
+        [Fact]
+        public async void GetNoExpirationTest()
+        {
+            var url = SignedUrlUtils.Create(
+                _fixture.ReadBucket,
+                _fixture.SmallObject,
+                _credentialsFilePath,
+                expiration: null);
+
+            // Verify that the URL works.
+            var response = await _httpClient.GetAsync(url);
+            var result = await response.Content.ReadAsByteArrayAsync();
+            Assert.Equal(_fixture.SmallContent, result);
+        }
+
+        [Fact]
+        public async void GetWithCustomHeadersTest()
+        {
+            var duration = TimeSpan.FromSeconds(5);
+
+            Func<HttpRequestMessage> createRequest = () => new HttpRequestMessage()
```

Comment:
Given that you're using this pattern frequently, I think we could probably write a helper method that calls it once, retains the response, waits, calls it again and validate that that fails, then returns the original response. Could definitely come later though.